pkg/controller/master/backup/restore.go

```diff
@@ -220,11 +273,117 @@ func (h *RestoreHandler) PersistentVolumeClaimOnChange(key string, pvc *corev1.P
 		return nil, nil
 	}
 
+	volumeName, err := getVolumeName(pvc)
+	if err != nil {
+		return nil, err
+	}
+
+	volume, err := h.volumeCache.Get(util.LonghornSystemNamespaceName, volumeName)
+	if err != nil {
+		return nil, nil
+	}
+
+	volumeCopy := volume.DeepCopy()
+	if volumeCopy.Annotations == nil {
+		volumeCopy.Annotations = make(map[string]string)
+	}
+
+	volumeCopy.Annotations[pvcNameSpaceAnnotation] = pvc.Namespace
+	volumeCopy.Annotations[pvcNameAnnotation] = pvc.Name
+	volumeCopy.Annotations[restoreNameAnnotation] = restoreName
+
+	if !reflect.DeepEqual(volume, volumeCopy) {
+		if _, err := h.volumes.Update(volumeCopy); err != nil {
+			return nil, err
+		}
+	}
+
 	logrus.Debugf("handling PVC updating %s/%s", pvc.Namespace, pvc.Name)
 	h.restoreController.EnqueueAfter(pvc.Namespace, restoreName, 5*time.Second)
 	return nil, nil
 }
 
+func (h *RestoreHandler) resolveLHEngineRef(lhEngine *lhv1beta2.Engine) (*lhv1beta2.Volume, error) {
+	refs := lhEngine.GetOwnerReferences()
+	if len(refs) == 0 {
+		return nil, fmt.Errorf("engine %s missing volume owner", lhEngine.Name)
+	}
+
+	ref := refs[0]
+	if ref.Kind != lhv1beta2.SchemeGroupVersion.WithKind("Volume").Kind {
+		return nil, fmt.Errorf("engine %s with wrong owner kind %v", lhEngine.Name, ref.Kind)
+	}
+
+	volume, err := h.volumeCache.Get(util.LonghornSystemNamespaceName, ref.Name)
+	if err != nil {
+		return nil, fmt.Errorf("engine %s owner volume %v not exist", lhEngine.Name, ref.Name)
+	}
+
+	if volume.UID != ref.UID {
+		return nil, fmt.Errorf("engine %s owner volume uuid not not consistent expected %v get %v",
+			lhEngine.Name, ref.UID, volume.UID)
+	}
+
+	return volume, nil
+}
+
+func (h *RestoreHandler) LHEngineOnChange(key string, lhEngine *lhv1beta2.Engine) (*lhv1beta2.Engine, error) {
+	if lhEngine == nil || lhEngine.DeletionTimestamp != nil || len(lhEngine.Status.RestoreStatus) == 0 {
+		return nil, nil
+	}
+
+	volume, err := h.resolveLHEngineRef(lhEngine)
+	if err != nil {
+		return nil, err
+	}
+
+	pvcNamespace, ok := volume.Annotations[pvcNameSpaceAnnotation]
+	if !ok {
+		return nil, nil
+	}
+
+	pvcName, ok := volume.Annotations[pvcNameAnnotation]
+	if !ok {
+		return nil, nil
+	}
+
+	restoreName, ok := volume.Annotations[restoreNameAnnotation]
+	if !ok {
+		return nil, nil
+	}
+
+	//vmRestore could be deleted after restoere completed
+	vmRestore, err := h.restoreCache.Get(pvcNamespace, restoreName)
```

Comment:
We can add a new indexer, so we don't need to save data in volume annotations.

In `pkg/indexeres/indexer.go`:

```go
const (
    // ...
    VMRestoreByPvcNameIndex            = "harvesterhci.io/vmrestore-by-pvc-name"
)

func Setup(ctx context.Context, server *server.Server, controllers *server.Controllers, options config.Options) error {
    // ...
    vmRestoreInformer := management.HarvesterFactory.Harvesterhci().V1beta1().VirtualMachineRestore().Cache()
	vmRestoreInformer.AddIndexer(VMRestoreByPvcNameIndex, VMRestoreByPvcName)
}

func VMRestoreByPvcName(obj *harvesterv1.VirtualMachineRestore) ([]string, error) {
	if obj.Status == nil {
		return []string{}, nil
	}
	result := []string{}
	for _, vr := range obj.Status.VolumeRestores {
		result = append(result, vr.PersistentVolumeClaim.ObjectMeta.Name)
	}
	return result, nil
}
```

In `pkg/controller/master/backup/restore.go`:
```go
func (h *RestoreHandler) LHEngineOnChange(key string, lhEngine *lhv1beta2.Engine) (*lhv1beta2.Engine, error) {
	if lhEngine == nil || lhEngine.DeletionTimestamp != nil || len(lhEngine.Status.RestoreStatus) == 0 {
		return nil, nil
	}

	volume, err := h.volumeCache.Get(util.LonghornSystemNamespaceName, lhEngine.Spec.VolumeName)
	if err != nil {
		return nil, err
	}

    pvcNamespace, pvcName := volume.Status.KubernetesStatus.Namespace, volume.Status.KubernetesStatus.PVCName

    vmRestores, err := h.restoreCache.GetByIndex(indexeres.VMRestoreByPvcNameIndex, pvcName)

    // ...
}
```

Reply:
As mentioned in https://github.com/harvester/harvester/pull/4869#discussion_r1431044095 and https://github.com/harvester/harvester/pull/4869#discussion_r1431052480, because the engine/replica restoring progress happens before pvc connected to LH volume, the annotation in LH volume `pvcNameSpaceAnnotation = "pvc.harvesterhci.io/namespace"` and `pvcNameAnnotation  = "pvc.harvesterhci.io/name"` are unavoidable.

If my understanding is correct, using indexers could avoid using LH volume annotation `restoreNameAnnotation  = "restore.harvesterhci.io/name"` to find matched VMRestore, however, this annotation already exists in the current master-head branch, and it is also used by pvc and VM.

I'd like to keep it for simpilcity, WDYT?

Reply:
Yes, it makes sense. Thank you.